.github/workflows/tests.yml

```diff
@@ -0,0 +1,404 @@
+              OPEN_RUNTIMES_START_COMMAND: "java -jar src/function/java-runtime-1.0.0.jar"
+            - runtimes: cpp-17
+              RUNTIME: cpp-17
+              TEST_CLASS: Base
+              ENTRYPOINT: tests.cc
+              IMAGE: openruntimes/cpp:v3-17
+              ARCH: linux/amd64,linux/arm/v6,linux/arm/v7,linux/arm64
+              TEST_SCRIPT: tests.sh
+              OPEN_RUNTIMES_BUILD_COMMAND: ""
+              OPEN_RUNTIMES_START_COMMAND: "src/function/cpp_runtime"
+            - runtimes: cpp-20
+              RUNTIME: cpp-20
+              TEST_CLASS: Base
+              ENTRYPOINT: tests.cc
+              IMAGE: openruntimes/cpp:v3-20
+              ARCH: linux/amd64,linux/arm/v6,linux/arm/v7,linux/arm64
+              TEST_SCRIPT: tests.sh
+              OPEN_RUNTIMES_BUILD_COMMAND: ""
+              OPEN_RUNTIMES_START_COMMAND: "src/function/cpp_runtime"
+            - runtimes: dotnet-7.0
+              RUNTIME: dotnet-7.0
+              TEST_CLASS: Base
+              ENTRYPOINT: Tests.cs
+              IMAGE: openruntimes/dotnet:v3-7.0
+              ARCH: linux/amd64,linux/arm64
+              TEST_SCRIPT: tests.sh
+              OPEN_RUNTIMES_BUILD_COMMAND: ""
+              OPEN_RUNTIMES_START_COMMAND: "dotnet src/function/DotNetRuntime.dll"
+            - runtimes: dotnet-6.0
+              RUNTIME: dotnet-6.0
+              TEST_CLASS: Base
+              ENTRYPOINT: Tests.cs
+              IMAGE: openruntimes/dotnet:v3-6.0
+              ARCH: linux/amd64,linux/arm64
+              TEST_SCRIPT: tests.sh
+              OPEN_RUNTIMES_BUILD_COMMAND: ""
+              OPEN_RUNTIMES_START_COMMAND: "dotnet src/function/DotNetRuntime.dll"
+    runs-on: ubuntu-20.04
+    name: ${{ matrix.runtimes }}
+    steps: 
+        - name: Clone Repository
+          uses: actions/checkout@v3.5.3
+        - name: Setup Docker Buildx
+          uses: docker/setup-buildx-action@v2.9.1
+        - name: Check Docker Buildx
+          run: docker buildx version
+        - name: Install PHP
+          uses: shivammathur/setup-php@2.25.4
+          with:
+            php-version: '8.0'
+        - name: Check PHP Version
+          run: php -v
+        - name: Install Dependencies
+          run: composer install
+        - name: Run Tests
+          run : |
+           RUNTIME="${{matrix.RUNTIME}}" ENTRYPOINT="${{ matrix.ENTRYPOINT }}" TEST_CLASS="${{ matrix.TEST_CLASS }}" OPEN_RUNTIMES_BUILD_COMMAND="${{ matrix.OPEN_RUNTIMES_BUILD_COMMAND }}" OPEN_RUNTIMES_START_COMMAND="${{ matrix.OPEN_RUNTIMES_START_COMMAND }}" sh ./tests.sh
+        - name: Login to DockerHub
+          uses: docker/login-action@v2.2.0
+          with:
+            username: ${{ secrets.DOCKER_USERNAME }}
+            password: ${{ secrets.DOCKER_PASSWORD }}
+        - name: Deploy to DockerHub
+          run: |
+           ARCH="${{ matrix.ARCH }}" IMAGE="${{ matrix.IMAGE }}" RUNTIME="${{ matrix.RUNTIME }}" sh ./deploy.sh
```

Comment:
Make sure the deploy step only runs when we create a tag. And the tests run on every push.

Checkout the travis file to see how we handle the deploy stage
https://github.com/open-runtimes/open-runtimes/blob/main/.travis.yml#L411-L412

Reply:
If possible, let's make tests run on deploy too. Tests first, then deploy. So when tests fail for some reason at that stage, it wont deploy.

Reply:
@Meldiron whenever a new release with a tag(version) is released all the steps run. It starts with the tests first if they fail it will suspend all the other steps and never execute the deployment step at all.